Status overhaul

milljm opened the pull request:
Refactor of the status bucket system, as well as the QueueManager (RunPBS).

In order to preserve continued PBS functionality, these two features needed to be refactored at the same time.

The status system is a fundamental change, and I fully expect failures from other apps. The common error we will see is:
```
FAILED (deprecated status bucket)
```
This is a simple fix, and will require modifying the offending tester to adjust the positional args of the status, and status message:
```python
    tester.setStatus("why we failed", self.bucket_failed)

# to:

    tester.setStatus(self.fail, "why we failed")

# or, if no message is required (hence the change to this more simple design):
    tester.setStatus(self.fail)
```

Commit:
Refactor the entire status bucket system

Very much a WIP. QueueManager/PBS functionality has not even
been looked at yet. With the talk the group had earlier about
introducing a new TestHarness ability to run a single 'test'
file within the PBS/Qsub script, the QueueManger/PBS thingy,
may be a thing of the past.

Refs #10456

Commit:
Remove current PBS implementation

Remove the current implementation of PBS and start work on
the new version.

Commit:
Adding QueueManager and RunPBS schedulers

Changes necessary to allow third party schedulers (only PBS at this time)
to launch and process jobs.

Refs #10456

Commit:
Remove deprecated Queueing paramaters

QueueManager no longer requires special 'tricks' to make
tests work within a third party scheduler.

Refs #10456

Commit:
Satisfy PyFlakes issues

Refs #10456

Commit:
More PyFlakes fixes

Refs #10456

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/178933/) on 716bb51 wanted to post the following:

View the site [here](http://mooseframework.inl.gov/docs/PRs/11312/site)

This comment will be updated on new commits.

Commit:
Allow deprecated statuses to be converted

Refs #10456

milljm commented on python/TestHarness/testers/Tester.py:
```diff
@@ -452,44 +386,37 @@ def checkRunnableBase(self, options):
             status = result_key.get(self.getTestName(), {}).get('FAIL', '')
             if not status:
-                self.setStatus('not failed', self.bucket_silent)
+                self.setStatus(self.bucket_silent)
```
fix this

milljm commented on python/TestHarness/schedulers/QueueManager.py:
```diff
@@ -25,565 +36,322 @@ def validParams():
+            self._finishAllJobs(job_data)
+            is_ready = False
+
+        return is_ready
```
refactor this method a bit more. if elif elif, or just return False within each to same cpu cycles.

milljm commented on python/TestHarness/schedulers/Scheduler.py:
```diff
@@ -449,217 +271,160 @@ def satisfyLoad(self):
+                # All done
+                with j_lock:
+                    job.setStatus(job.finished)
+                    #print 'RUNNER FINISHED JOB: %s' % (job.getTestName())
```
remove debug print

Commit:
Add ability to detect invalid output

Some tests are generating invalid unicode characters.
Instead of crashing, set the specific test to a failed state,
so those tests can be fixed.

Refs #10456

Commit:
Address comments

Remove debug print.
Refactor isProcessReady.
Fix Tester silent bucket usage.

Refs #10456

Commit:
Do not modify base_dir. Update helper script

Do not modify the harness.base_dir variable once its been set
by findtestroot. This breaks some applications that do have it
set.

Update the `are_my_queued_jobs_finished.py` script to handle
the new results file format.

Refs #10456

Commit:
Fix bad output

Instead of ignoring bad output, we need to convert it
to something that will not break civet.

If not using --spec-file, then use base_dir rather than
trusting we are still in the proper directory.

Refs #10456

Commit:
Convert null characters to something else

We can not print null characters to stdout. This seems to break
civet (or rather breaks postgress when civet saves data).

Refs #10456

Commit:
Clean up QueueManager a bit

Refactor a createTestStatus method based on incoming session
file.

isProcessReady needed a little work. This method was by far the
most troublesome. When do we actually process a tester instead of
ignoring it, and launching it? Was it silent previously? Deleted?
How about now? Did the user supply different arguments that
would now cause the tester to run? In the end, I believe the best
solution was just making the tester equal the same status it was
during the previous launch. The only caveat being, to print a
nice 'QUEUED' message instead of 'NA' for testers we know have
launched, yet currently have no results to display.

Refs #10456

milljm commented:
Mammoth is failing due to that application outputting bad unicode characters. This is now an issue, because we need to save output when performing PBS operations (output is stored in json format).

permcody commented:
I've opened an issue for @YaqiWang to look at this. It's there because of some "bad practice" debugging files that are being written. Those should just be removed or use some other form of logging that's more robust. We can serialize these tests if we need this to go in sooner.

milljm commented:
I have additional work I want to add to this PR

Commit:
Fix skipped/deleted/silent tests running (or not)

Clean up the logic when deciding whether or not to handle test _not_
submitted to the third party queue. It gets tricksy! Especially when
a user can alter the arguments being passed.

Allow 'addCaveats' to support lists and strings. It didn't make sense
that getCaveats returned a list, but addCaveats couldn't handle it.

Refactored formatStatusMessage into Util.py where it arguably belongs
with the other formatting methods. Admittedly, this probably belongs
in another PR.

Refs #10456

Commit:
Check toplevel value is a dictionary

I am saving additional key/values into the session file which dictates how
the TestHarness operates. Unfortunately these values are not dictionaries.
Which is what 'are_queued_jobs_finished' was expecting.

This is a simple fix for now, and will be addressed again in future PRs when
we move the results_storage methods into its own class (I dislike passing
results through the argparse namespace).

Refs #10456

milljm commented:
I found an issue with MAX FAILURES. Investigating...

Commit:
Smarter status checks

Refs #10456

Commit:
Fix MAX FAILURES detection

Faulty logic in the way silent deleted tests were being counted
as a failure.

Refs #10456

Commit:
Allow tests to pass with invalid characters

Instead of failing tests which produce invalid characters in their
output, add a caveat as to the issue, and all the test to pass.

Refs #10456

permcody commented:
@idaholab/moose-team - Jason has been working on Refactoring his own work for quite awhile. This PR is too big to review thoroughly and too critical to just merge. If anyone wants to pull this branch and play around with it, that would be very valuable.
